warp/aggregator/aggregator.go

```diff
@@ -36,16 +70,134 @@ func (a *Aggregator) AggregateSignatures(ctx context.Context, unsignedMessage *a
 	if err != nil {
 		return nil, err
 	}
-	job := newSignatureAggregationJob(
-		a.client,
-		pChainHeight,
-		a.subnetID,
-		quorumNum,
-		quorumNum,
-		params.WarpQuorumDenominator,
-		a.state,
-		unsignedMessage,
+
+	log.Debug("Fetching signature",
+		"a.subnetID", a.subnetID,
+		"height", pChainHeight,
 	)
+	validators, totalWeight, err := avalancheWarp.GetCanonicalValidatorSet(ctx, a.state, pChainHeight, a.subnetID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get validator set: %w", err)
+	}
+	if len(validators) == 0 {
+		return nil, fmt.Errorf("%w (SubnetID: %s, Height: %d)", errNoValidators, a.subnetID, pChainHeight)
+	}
+
+	return aggregateSignatures(ctx, a.client, unsignedMessage, validators, quorumNum, totalWeight)
+}
+
+func aggregateSignatures(
+	ctx context.Context,
+	client SignatureGetter,
+	unsignedMessage *avalancheWarp.UnsignedMessage,
+	validators []*avalancheWarp.Validator,
+	quorumNum uint64,
+	totalWeight uint64,
+) (*AggregateSignatureResult, error) {
+	var (
+		// [signatureLock] must be held when accessing [blsSignatures],
+		// [signersBitset], or [signatureWeight] in the goroutine below.
+		signatureLock   = sync.Mutex{}
+		signatures      = make([]*bls.Signature, 0, len(validators))
+		signersBitset   = set.NewBits()
+		signatureWeight = uint64(0)
+	)
+
+	// Create a child context to cancel signature fetching if we reach [maxNeededQuorumNum] threshold
+	signatureFetchCtx, signatureFetchCancel := context.WithCancel(ctx)
+	defer signatureFetchCancel()
+
+	wg := sync.WaitGroup{}
+	wg.Add(len(validators))
+	for i, validator := range validators {
+		var (
+			i         = i
+			validator = validator
+			// TODO: update from a single nodeID to the original slice and use extra nodeIDs as backup.
+			nodeID = validator.NodeIDs[0]
+		)
+		go func() {
+			defer wg.Done()
+
+			log.Debug("Fetching warp signature",
+				"nodeID", nodeID,
+				"index", i,
+			)
+
+			signature, err := client.GetSignature(signatureFetchCtx, nodeID, unsignedMessage)
+			if err != nil {
+				log.Debug("Failed to fetch warp signature",
+					"nodeID", nodeID,
+					"index", i,
+					"err", err,
+				)
+				return
+			}
+			log.Debug("Retrieved warp signature",
+				"nodeID", nodeID,
+				"index", i,
+				"signature", hexutil.Bytes(bls.SignatureToBytes(signature)),
```

Comment:
can we assign this signature to some variable before calling this so we can use it below as well?